Reorganize docs (collections, examples)

koverholt opened the pull request:
Addresses some documentation reorganization tasks from #915:

* No content changes, only reorganization
* Make collections docs consistent and consolidated:
 * Overview
 * Example
 * Create
 * ...
 * API
* Move a couple of longer examples to `Examples` docs section
* Add `Getting Started` section to top of docs
* Move `Comparison to Blaze` out of `Dask Array` section
* Fix some docs build warnings

mrocklin commented:
General thoughts are that I really like the idea of adding predictable structure around the collections documents.  However, doing so highlights a lot of flaws in our content.  This particular reorganizations highlights some parts of the documentation a bit more than is probably warranted.  Small examples:

* The dask.array example is very developer focused.  I'm not sure we should present it to users as the main example.
* Ghost computations are fringe enough that I don't think they should be in the main narrative of dask.arrays, the same goes for random.  
*  The dask.bag example exists in two places

However, improving the content is probably my job.  I'll take a crack at this later today.

koverholt commented:
Oops, I forgot to deduplicate the Dask Bag example. Fixed.

I had the same thoughts as you regarding ghost cells and random. As a first cut, I've put them in `Array Creation`, which seems a bit long as-is. I also considered moving them to an advanced section or the like.

mrocklin commented:
Rather than an advanced section I'm still in favor of having many small documents.  It's much easier to write this sort of thing and, in the case of self contained topics like ghosting, probably matches what readers want as well.  How about a standard set of documents for each collection as you have here (overview, examples, api, etc.) and then a free form list of documents following.

koverholt commented:
Yes, that document structure sounds reasonable (overview, examples, api, etc.). I can take a second cut later today.

koverholt commented:
I took another cut at organizing the docs structure under each collection, and I reworked the organization of the examples (and added a few more simple examples).

mrocklin commented:
Great, I'll take a pass through.

mrocklin commented on docs/source/array-creation.rst:
````````diff
@@ -109,26 +109,103 @@ A good choice of ``chunks`` follows the following rules:
-Example
-```````
+Slicing
+-------
````````
Why is slicing in array creation?

koverholt replied:
See below comment regarding stack and concatenate.

mrocklin commented on docs/source/array-creation.rst:
```diff
@@ -109,26 +109,103 @@ A good choice of ``chunks`` follows the following rules:
+
+
+Stack and Concatenate
+---------------------
```
Same with stack and concatenate.  These seem like separate topics from creating dask arrays.  I'm more in favor of keeping these as separate files as before.  I think it makes them more discoverable.

koverholt replied:
My attempt to consolidate these topics seems to have failed. The docs for Dask arrays were a bit disparate at first glance compared to the other collections. I separated them back out to single pages for the time being.

mrocklin replied:
Dask array is just much more complex than the other collections, which are generally straightforward.  There is no way to include all of the topics in a linear narrative.  I think that treating the documentation for dask.array as small narrative + extensive reference pages is the right approach.

koverholt replied:
Makes sense. Thanks for iterating through that with me in this PR.

mrocklin commented on docs/source/array-examples.rst:
```diff
@@ -0,0 +1,9 @@
+
+   examples/array-random.rst
+   examples/array-numpy.rst
+   examples/array-pickle.rst
```
What are our objectives for these examples?  Something they can run immediately?  Something that shows off how they would use dask.array in the real world?

koverholt replied:
I often go into docs looking for short examples I can experiment with to grok the concepts of a library, or to use as a starting point to build from. I find it convenient when they are centralized and easily accessible. Sometimes notebooks and blog posts go stale. These examples could also be tested as doctests if desired.

mrocklin replied:
I agree with these objectives.  I think I'm against having them be as prominent in the documents until they're more built out.  What's up here right now feels simplistic.  I would feel more excited about this section if it went over common problems that people actually have, rather than small and artificial examples.  

However, again, content generation like this is probably something that I should handle.  If there are other sections that seem like they're good starts that should be built out then please bring them to my attention.  I'll put this on my todo list for early next week.

koverholt replied:
Sounds good. I may have bit off more than I meant to chew in this PR by reorganizing many different things, but hopefully this centralizes things for new folks coming to Dask, even if they are short-term placeholder examples. I'm happy to keep iterating on and improving the examples.

mrocklin commented on docs/source/array-random.rst:
```diff
@@ -1,5 +1,5 @@
+=============
 
 ``dask.array`` copies the ``numpy.random`` module for all univariate
 distributions.  The interface to each function is identical except for the
```
This file can probably be deleted and just replaced with nicer API docs

mrocklin replied:
In the api file that it

koverholt replied:
I had the same thought about this page (consolidating into API docs). Could you handle this or suggest the best way to do this? My Sphinx knowledge here is limited.

mrocklin replied:
I've used a bit of copy-paste and vi magic to populate array-api.rst explicitly

koverholt replied:
This is nice. I did the same for the `bag` and `dataframe` API pages. I removed the `array-random` subpage references in this thread in lieu of the updated API pages.

mrocklin commented on docs/source/bag-api.rst:
```diff
@@ -0,0 +1,32 @@
+-----------
+
+.. autoclass:: Bag
+   :members:
```
We might want to do a full autosummary of Bag up at the top.  I suspect that this will make methods more discoverable as well as give a quick at-a-glance summary of available functionality.

```
.. autosummary::
   Bag
   Bag.map
   Bag.filter
   ...
```

koverholt replied:
I think this is really helpful. I manually added a full autosummary at the top with the public members of Bag. But there [doesn't seem to be an automatic way to do this with autosummary](http://stackoverflow.com/questions/20569011/python-sphinx-autosummary-automated-listing-of-member-functions) other than extending it manually.

mrocklin commented on docs/source/bag-overview.rst:
```diff
@@ -0,0 +1,31 @@
+There can be repeated elements but you can't ask for a particular element.
+
+.. _multiset: http://en.wikipedia.org/wiki/Bag_(mathematics)
+
```
We've stripped out almost all of the content here.  I'm not sure how best to arrange data in separate files instead of right here.

koverholt replied:
I moved the bag execution section back into the overview page. So now the sections are: Overview, Examples, Create Dask Bags, API. We can continue to iterate here along with the other collection sections.

mrocklin commented on docs/source/blaze.rst:
```diff
@@ -1,5 +1,5 @@
-Blaze
-=====
+Comparison to Blaze
+===================
```
I think that we should remove this file

koverholt replied:
Removed.

mrocklin commented on docs/source/examples-tutorials.rst:
```diff
@@ -0,0 +1,31 @@
+Examples and Tutorials
+======================
+
```
Should probably link to the `dask-tutorial` repository front-and-center

koverholt replied:
Added a link to the `dask-tutorial` repository.

mrocklin commented on docs/source/examples/array-numpy.rst:
```diff
@@ -0,0 +1,13 @@
+    >>> x = np.arange(1000)
+    >>> y = da.from_array(x, chunks=(100))
+    >>> y.mean().compute()
+    499.5
```
I'm concerned that people will actually do this, which would be a mistake.  If we have this example we should include a pretty strong disclaimer along with it saying that this is a bad idea.

koverholt replied:
I added some trivial examples to get started here (since the only existing array example was the pickle files. Happy to change this to a more realistic example with NumPy or remove it for now.

mrocklin replied:
How about text like the following:

We can create dask arrays from any object that implements NumPy slicing, like a `numpy.ndarray` or on-disk formats like `h5py` or `netCDF` `Dataset` objects.  This is particularly useful with on disk arrays that don't fit in memory but, for simplicity's sake, we show how this works on a NumPy array.  This isn't particularly valuable (the NumPy array already works in memory just fine) but is easy to play with.

My objective here is to avoid naive users changing all of their perfectly fine NumPy code with dask.array code to no benefit.  We run across these cases occasionally, hence my aversion.

koverholt replied:
Updated the language here.

mrocklin commented on docs/source/examples/array-pickle.rst:
```diff
@@ -0,0 +1,23 @@
+
+    chunks = ((1000, 1000), (1000, 1000, 1000))
+
+    x = da.Array(dsk, 'x', chunks)
```
This example should be removed from any place where every-day users would see it.  This is really only useful for developers.

koverholt replied:
I moved this back to its original location (bottom of create arrays page).

mrocklin commented on docs/source/examples/array-random.rst:
```diff
@@ -0,0 +1,11 @@
+Creating random arrays
+======================
+
+We can create arrays with random variables using ``da.random.normal``.
```
Suggest the following language:

> In a simple case we create arrays with random data using the ``da.random`` module.

For the following reasons

1.  The term random variable is a bit special and doesn't apply here.  Instead we have normally distributed random data
2.  We don't give the impression that you have to use `da.random.normal` to create random data, but rather there is a full module to choose from

koverholt replied:
I updated the language here.

mrocklin commented on docs/source/examples/bag-json.rst:
```diff
@@ -0,0 +1,17 @@
+Simple example
+==============
```
Maybe a more descriptive title like `Read JSON records from disk`

koverholt replied:
Updated; I forgot to update these placeholder titles.

mrocklin commented on docs/source/examples/dataframe-csv.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. code-block:: python
+
+    >>> import dask.dataframe as ddf
```
The standard import pronoun is `dd`

    import dask.dataframe as dd

koverholt replied:
Fixed.

mrocklin commented on docs/source/examples/imperative-array.rst:
```diff
@@ -0,0 +1,49 @@
+Simple parallel example
+=======================
```
Perhaps a more informative title.  This ends up populating TOC and (maybe) sidebars

koverholt replied:
Updated; I forgot to update these placeholder titles.

mrocklin commented on docs/source/imperative-api.rst:
```diff
@@ -0,0 +1,8 @@
+API
+===
+
+.. currentmodule:: dask.imperative
```
Suggest `autosummary`

koverholt replied:
Fixed.

mrocklin commented on docs/source/examples/array-random.rst:
```diff
@@ -0,0 +1,12 @@
+Creating random arrays
+======================
+
+In a simple case, we can create arrays with random data using the `da.random`
```
Want double backticks for code highlighting.  `rst` differs from `markdown`

koverholt replied:
Fixed.

mrocklin commented on docs/source/examples/array-random.rst:
```diff
@@ -0,0 +1,12 @@
+.. code-block:: python
+
+    >>> import dask.array as da
+    >>> x = da.random.normal(0, 1, size=(100000,100000), chunks=(10000, 10000))
```
A chunk size of `(1000, 1000)` is probably much more safe here.  A chunk of `(10000, 10000)` of `float64`s is 800 MB.  We often have 10-20 chunks floating about so this can become dangerous.  Shooting for chunks in the 1-100MB range is generally safer.

koverholt replied:
Fixed the chunk size.

mrocklin commented:
I've gone through and updated the collections docs.  Commit-by-commit review is probably best.

koverholt commented on docs/source/dataframe-create.rst:
```diff
@@ -58,22 +46,43 @@ In particular column access on a dask.dataframe backed by a ``bcolz.ctable``
    >>> c = Castra(path='/my/castra/file')
    >>> df = c.to_dask()
 
+.. _Castra: http://github.com/Blosc/castra
```
This seems to live at `https://github.com/blaze/castra` (but currently redirects).

mrocklin replied:
Fixed

koverholt commented on docs/source/dataframe-create.rst:
```diff
@@ -58,22 +46,43 @@ In particular column access on a dask.dataframe backed by a ``bcolz.ctable``
+You can create a plan to arrange many Pandas frames into a sequence with normal
+for loops using :doc:`dask.delayed<imperative-overview>` and then convert these
+into a dask dataframe later.  See :doc:`documentation on using dask.delayed with
+collections<imperative-collections>`.
```
What's the use case of using `dask.delayed` instead of `Bag.to_dataframe` here? If you're already using `dask.delayed` elsewhere? It's not immediately clear to me.

mrocklin replied:
Added an notebook of the feather example in the next commit.  Bag is good when you can `map` your way to victory.  Delayed is good when your problem is more complex.

koverholt commented on docs/source/dataframe-overview.rst:
```diff
@@ -76,19 +80,51 @@ What definitely works?
+
+.. code-block:: python
+
+   >>> dask.set_options(get=dask.multiprocessing.get)
```
Since this example only shows the multiprocessing scheduler, maybe a link to the scheduler docs page?

mrocklin replied:
Fixed

koverholt commented on docs/source/dataframe-overview.rst:
```diff
@@ -1,25 +1,42 @@
+   :width: 50%
+
+Dask dataframes coordinate many Pandas arrays arranged along an index.  These
+dataframes may live on disk or on other machines.
```
> These dataframes may live on disk or on other machines.

Is this referring to the serialized dataframes on disk, or after being loaded as Dask dataframes, or...?

mrocklin replied:
Is this more clear if we change it to "These *Pandas* dataframes may live ..."

koverholt commented on docs/source/faq.rst:
```diff
@@ -99,5 +99,14 @@ Frequently Asked Questions
+serialization is necessary.  When operating with the distributed memory or
+multiprocessing scheduler Dask uses cloudpickle_ to serialize functions to send
+to worker processes.  cloudpickle supports almost any kind of function,
+including lambdas, closures, partials and functions defined interactively.
```
It is worth mentioning here how one might debug/troubleshoot a function that isn't serializable via cloudpickle or test whether it is? Or somewhere else (e.g., diagnostics/troubleshooting page)?

mrocklin replied:
Added another paragraph on common issues and how to verify serializability.

koverholt commented on docs/source/array-creation.rst:
```diff
@@ -13,9 +13,10 @@ their arrangement can be as simple or as complex as the situation dictates.
-For example the HDF5 file format has the ``h5py`` project, which provides a
-``Dataset`` object into which we can slice in NumPy fashion
+Many storage formats have Python projects that expose storage using NumPy
+slicing syntax.  These include HDF5, NetCDF, BColz, Zarr, GRIB, etc..  For
```
Double period in `etc..`

mrocklin replied:
I believe that double period on etc. is common when ending a sentence with etc..

koverholt commented on docs/source/array-creation.rst:
```diff
@@ -13,9 +13,10 @@ their arrangement can be as simple or as complex as the situation dictates.
+Many storage formats have Python projects that expose storage using NumPy
+slicing syntax.  These include HDF5, NetCDF, BColz, Zarr, GRIB, etc..  For
+example the ``HDF5`` file format has the ``h5py`` Python project, which
+provides a ``Dataset`` object into which we can slice in NumPy fashion
```
Missing period.

mrocklin replied:
Fixed

koverholt commented on docs/source/array-creation.rst:
```diff
@@ -132,3 +148,73 @@ As an example, we might load a grid of pickle files known to contain 1000 by
+---------
+
+If you have a small amount of data, you can call ``np.array`` on your dask array to tur
+in to a normal NumPy array:
```
Typo in `tur in to a normal`

mrocklin replied:
Fixed

koverholt commented on docs/source/array-creation.rst:
```diff
@@ -132,3 +148,73 @@ As an example, we might load a grid of pickle files known to contain 1000 by
+
+   >>> x = dataset[1000:2000, :2000]  # pull out numpy array from on-disk object
+
+This API has become a standard in Scientific Python.  Dask works with any
```
Maybe use something other than `Scientific Python` to avoid confusion with SciPy, or the [ScientificPython](http://dirac.cnrs-orleans.fr/plone/software/scientificpython/) library?

mrocklin replied:
Changed to numeric Python ecosystem

koverholt commented on docs/source/array-overview.rst:
```diff
@@ -6,10 +6,25 @@ algorithms, cutting up the large array into many small arrays. This lets us
+   :align: right
+
+Dask arrays coordinate many NumPy arrays arranged into a grid.  These
+arrays may live on disk or on other machines.
```
Same comment as the dataframe docs:

> These arrays may live on disk or on other machines.

Is this referring to the serialized arrays on disk, or after being loaded as Dask arrays, or...?

mrocklin replied:
Inserted "NumPy"

koverholt commented on docs/source/dataframe-create.rst:
```diff
@@ -1,40 +1,28 @@
 Create Dask DataFrames
-----------------------
+======================
```
Is it worth adding a `Storing Dask Dataframes` subpage?

mrocklin replied:
Maybe.  I'm not opinionated on this one way or the other.

koverholt commented:
This looks great! I left some minor comments.

koverholt commented on docs/source/bag-creation.rst:
```diff
@@ -65,3 +65,53 @@ attached directly to bags with ``.str.methodname``:
+``db.from_imperative``
+----------------------
+
+You can construct a dask bag from :doc:`dask.delayed<imperative.rst>` values
```
Broken reference, `dask.delayed<imperative.rst>` should be `dask-delayed <imperative>` (with space).

koverholt commented on docs/source/bag-creation.rst:
```diff
@@ -65,3 +65,53 @@ attached directly to bags with ``.str.methodname``:
+
+You can construct a dask bag from :doc:`dask.delayed<imperative.rst>` values
+using the ``db.from_imperative`` function.  See
+:doc:`documentation on using dask.delayed with collections`<imperative-collections>`
```
Space between `collections` and `<`

koverholt commented on docs/source/faq.rst:
```diff
@@ -101,11 +101,25 @@ Frequently Asked Questions
+    serialize these objects (or functions that implicitly rely on these
+    objects) will result in scheduler errors.  You can verify that your objects
+    are easily serializable by running them through the
+    ``cloudpickle.dumps/loads`` functions
```
Did you mean to indent these paragraphs to block-quote level?

koverholt commented:
After this latest update, this looks great! Thanks again for your efforts and iterating on this with me, it was a good Dask learning experience on my end.

mrocklin commented:
This looks good to me.  Merging soon if no comments.
